tt_metal/tt_stl/any_range.hpp

```diff
@@ -0,0 +1,887 @@
+
+    [[nodiscard]] const type_info *id() const override {
+        static const type_info unique{};
+        return &unique;
+    }
+
+    void uninitialized_copy_to(void *dst_ptr) const override {
+        ::new (dst_ptr) final_iterator_adaptor_type(static_cast<const final_iterator_adaptor_type &>(*this));
+    }
+
+    void uninitialized_move_to(void *dst_ptr) noexcept override {
+        ::new (dst_ptr) final_iterator_adaptor_type(static_cast<final_iterator_adaptor_type &&>(*this));
+    }
+
+    [[nodiscard]] bool operator==(
+        const IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag> &other) const override {
+        return id() == other.id() and iterator == static_cast<const AnyIteratorAdaptor &>(other).iterator;
+    }
+
+    [[nodiscard]] typename TAnyIterator::reference operator*() const override {
+        return static_cast<typename TAnyIterator::reference>(*iterator);
+    }
+
+    [[nodiscard]] typename TAnyIterator::pointer operator->() const override { return detail::to_address(iterator); }
+
+    void operator++() override { ++iterator; }
+};
+
+template <class TAnyIterator, class TIterator>
+class AnyIteratorAdaptor<TAnyIterator, TIterator, std::forward_iterator_tag>
+    : public AnyIteratorAdaptor<TAnyIterator, TIterator, std::input_iterator_tag> {
+   public:
+    using AnyIteratorAdaptor<TAnyIterator, TIterator, std::input_iterator_tag>::AnyIteratorAdaptor;
+
+    [[nodiscard]] TAnyIterator operator++(int) override { return TAnyIterator(this->iterator++); }
+};
+
+template <class TAnyIterator, class TIterator>
+class AnyIteratorAdaptor<TAnyIterator, TIterator, std::bidirectional_iterator_tag>
+    : public AnyIteratorAdaptor<TAnyIterator, TIterator, std::forward_iterator_tag> {
+   public:
+    using AnyIteratorAdaptor<TAnyIterator, TIterator, std::forward_iterator_tag>::AnyIteratorAdaptor;
+
+    void operator--() override { --this->iterator; }
+
+    [[nodiscard]] TAnyIterator operator--(int) override { return TAnyIterator(this->iterator--); }
+};
+
+template <class TAnyIterator, class TIterator>
+class AnyIteratorAdaptor<TAnyIterator, TIterator, std::random_access_iterator_tag>
+    : public AnyIteratorAdaptor<TAnyIterator, TIterator, std::bidirectional_iterator_tag> {
+   public:
+    using AnyIteratorAdaptor<TAnyIterator, TIterator, std::bidirectional_iterator_tag>::AnyIteratorAdaptor;
+
+    void operator+=(typename TAnyIterator::difference_type offset) override { this->iterator += offset; }
+
+    [[nodiscard]] TAnyIterator operator+(typename TAnyIterator::difference_type offset) const override {
+        return TAnyIterator(this->iterator + offset);
+    }
+
+    [[nodiscard]] typename TAnyIterator::difference_type operator-(
+        const IAnyIteratorAdaptor<TAnyIterator> &other) const override {
+        return this->id() == other.id() ? this->iterator - static_cast<const AnyIteratorAdaptor &>(other).iterator
+                                        : this->iterator - TIterator{};
```

Comment:
Can you please clarify, what does it mean to subtract two iterators of a different type? Shouldn't we do TT_FATAL in this case?

Reply:
If I depend on those macros in a header then we need to publicly depend on fmt which has recently just been made a private dependency after some effort. Subtracting two iterators of different underlying types should be considered UB from the user's perspective, but this sanity check should at least prevent the compiler from making unpredictable optimizations in case this is ever done on accident.